### ISSUE TITLE
Bug: incorrect handling of minProperties and maxProperties in optional attribute

### ISSUE THREAD

**Issue by rtucek** [justinrainbow/json-schema#215]
Given schema:

```json
{
  "type": "object",
  "additionalProperties": false,
  "required": [
    "session",
    "include_hotel",
    "arrival",
    "departure",
    "ip",
    "travellers",
    "language_iso_code"
  ],
  "properties": {
    "session": {
      "type": "string"
    },
    "campaign": {
      "type": "integer"
    },
    "service_provider": {
      "type": "integer"
    },
    "include_hotel": {
      "type": "boolean"
    },
    "arrival": {
      "type": "string",
      "pattern": "^[0-9]{4}-[0-9]{2}-[0-9]{2}$"
    },
    "departure": {
      "type": "string",
      "pattern": "^[0-9]{4}-[0-9]{2}-[0-9]{2}$"
    },
    "ip": {
      "type": "string"
    },
    "offerId": {
      "type": "integer",
      "minimum": 1
    },
    "quiet": {
      "type": "boolean"
    },
    "language_iso_code": {
      "type": "string"
    },
    "travellers": {
      "type": "array",
      "minItems": 1,
      "maxItems": 16,
      "items": {
        "type": "integer"
      }
    },
    "rooms": {
      "type": "object",
      "additionalProperties": false,
      "minProperties": 1,
      "maxProperties": 2,
      "properties": {
        "size_1": {
          "type": "number"
        },
        "size_2": {
          "type": "number"
        }
      }
    },
    "offset": {
      "type": "integer"
    }
  }
}
```
And this data:
```json
{
  "campaign": 1,
  "offset": 0,
  "service_provider": 3,
  "include_hotel": false,
  "arrival": "2016-01-20",
  "departure": "2016-01-27",
  "session": "foo",
  "ip": "bar",
  "language_iso_code": "de",
  "travellers": [
    18,
    18,
    18
  ]
}
```

Proofed by http://jsonschemalint.com/draft4/ the json is valid.
However the validator is returning the following error:
```php
[
    0 =>  [
        "property" => "rooms",
        "message" => "Must contain no more than 2 properties",
        "constraint" => "maxProperties",
        "maxProperties" => 2
    ]
];
```

This should not happen because the property rooms is not given.
Thus their is no reason for a check on maxProperties constraint.

**Comment by bighappyface** [comment]
@rtucek good find. Pull requests welcome :wink:

**Comment by rtucek** [comment]
Here we got #224:
Fix + regression test + build passed.

**Comment by rtucek** [comment]
Ok, so after my little PR mess, I've provided 2 different approaches for fixing the issue.

#232 as you suggested
#233 as my original and restored proposal

Note that this time I haven't submitted a test.
Yes, I've noticed the json-schema/JSON-Schema-Test-Suite repo and I'm aware of that the phpunit is successfully testing against them.

So far, so good, but it's false-failing against my document which is perfectly valid, but I little more complex compared the standard tests (and that's how you trigger the bug in this library, having some nested constraints).

And this was the reason why I used this document as a regression test - just for having a pre-post comparison and for demonstrating the code change.
I'm sry that I'm failing simplifying the test any further without demonstrating anything xD
May you give me a concrete example of what you're thinking of?

**Comment by bighappyface** [comment]
Closed with #232
